src-test/org/etools/j1939_84/controllers/part11/Part11Step01ControllerTest.java

```diff
@@ -118,12 +126,95 @@ public void testGetTotalSteps() {
+        when(engineSpeedModule.getKeyState()).thenReturn(KEY_ON_ENGINE_RUNNING);
+        when(engineSpeedModule.getEngineSpeedAsString()).thenReturn("0.0 RPMs");
+
+        runTest();
+
+        verify(engineSpeedModule, atLeastOnce()).getEngineSpeedAsString();
+        verify(engineSpeedModule, atLeastOnce()).getKeyState();
+
+        String expectedMessages = "";
+        assertEquals(expectedMessages, listener.getMessages());
+
+        String expectedMilestones = "";
+        assertEquals(expectedMilestones, listener.getMilestones());
+
+        String expectedResults = "";
+        expectedResults += "Initial Engine Speed = 0.0 RPMs" + NL;
+        expectedResults += "Final Engine Speed = 0.0 RPMs" + NL;
+        assertEquals(expectedResults, listener.getResults());
+    }
+
+    @Test
+    public void testWaitForKeyOn() {
+        when(engineSpeedModule.getKeyState()).thenReturn(KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_OFF,
+                                                         KEY_ON_ENGINE_OFF,
+                                                         KEY_OFF,
+                                                         KEY_ON_ENGINE_RUNNING);
+        when(engineSpeedModule.getEngineSpeedAsString()).thenReturn("148.6 RPMs");
 
         runTest();
 
-        assertEquals("", listener.getMessages());
-        assertEquals("", listener.getResults());
+        verify(engineSpeedModule, atLeastOnce()).getKeyState();
+        verify(engineSpeedModule, atLeastOnce()).getEngineSpeedAsString();
+
+        verify(mockListener).onUrgentMessage("Please turn Key ON/Engine RUNNING", "Adjust Key Switch", WARNING);
+
+        String expectedMessages = "Waiting for Key ON/Engine RUNNING..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine RUNNING..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine RUNNING..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine RUNNING..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine RUNNING..." + NL;
+        expectedMessages += "Waiting for Key ON/Engine RUNNING...";
+        assertEquals(expectedMessages, listener.getMessages());
+
+        String expectedMilestones = "";
+        assertEquals(expectedMilestones, listener.getMilestones());
+
+        String expectedResults = "Initial Engine Speed = 148.6 RPMs" + NL;
+        expectedResults += "Final Engine Speed = 148.6 RPMs" + NL;
+        assertEquals(expectedResults, listener.getResults());
+    }
+
+    @Test
+    public void testEngineThrowInterruptedException() {
+
+        when(engineSpeedModule.getKeyState()).thenReturn(KEY_OFF);
+        when(engineSpeedModule.getEngineSpeedAsString()).thenReturn("300.0 RPMs");
+
+        new Timer().schedule(new TimerTask() {
```

Comment:
instance.stop() isn't really used in the app.  So, I have been switching to a pattern where we check the question listener.  We need to check at least one of the questions with an answer of NO or CANCLE to verify the question listener all the way through.  I have come across this two or three times now.